### ISSUE TITLE
rp2: crash and memory corruption with threading

### ISSUE THREAD

**Issue by robert-hh** [micropython/micropython#7124]
In a use case, where two threads are created and massive UART write happens, various kinds of crashes and memory corruptions occur. Test script:
```
from machine import UART, Pin
import time
import _thread

uart = UART(1, 115200 , parity=None, stop=1, bits=8, tx=Pin(4), rx=Pin(5), txbuf=32, timeout=10)

def send(command):
        write_buffer = bytearray(len(command) + 3)
        write_buffer[0: len(command)] = bytes(command, 'ASCII')
        write_buffer[-3:] = b'\xff\xff\xff'
        uart.write(write_buffer)

def burst():
    msg="0123456789" * 8 + "\r\n"
    n=0
    while True:
        uart.write("%d: " % n)
        uart.write(msg)
        n += 1

def count_count():
    n = 1
    while True:
        print("Count Count counts", n)
        n += 1
        time.sleep(1)

_thread.start_new_thread(burst, ())

# burst()
count_count()
```
The function send() is not called, but the behavior changes on whether is is present in the code or commented out. For the test, UART1 tx is connected with a UART/USB bridge to a PC showing the data stream. The following test set-ups were made:

- With send() code, thread 0 REPL, thread 1 burst(): 
         Result: memory allocation error at msg 5721, rare character doubling, like:
                "MemoryError: memory allocation failed, allocating 875706184 bytes"
                "5475: 011223344556677889900112233456789012345678901234567890123456789012345678901234567890123456789"
- With send() code, thread 0 count_count(), thread 1 burst(): 
        Result: Lock-up after msg 5721, rare char doubling
- Without send() code, thread 0 count_count(), thread 1 burst():
        Result: No lock-up, rare char doubling
- Without send() code, thread 0 REPL, thread 1 burst()
        Result: No lock-up, no transmission errors

- With send() code, thread 0 burst(), thread 1 count_count()
        Result: sending complete rubbish after msg 5720, like, and freezing both threads after a while:
                "43: h5��� 44: h5�� 45: h5�0� 46: h5�P� 47: h5�p� 48: h5��� 49: h5��� 50: h5��� 51: h5{�� 52: h5��" and so on
- Without send() code, thread 0 burst(), thread 1 count_count()
        Result No problem. stopped after msg 100_000

Using critical section instead of lock does not change the game (not surprising).
Increasing the thread stack size has no effect (expected).

Edit: Using the previous version of machine_uart.c without buffering fails too. Only the character doubling error seems to be gone. But since that was rare, the observation time may have been too short.

**Comment by dpgeorge** [comment]
My first guess would be that it's something to do with the GC interacting between the two cores/threads.  Can you try writing a test that does not do any memory allocations when running (ie only at start up)?

**Comment by robert-hh** [comment]
Good guess, indeed. Without allocation it looks much better. Same test, 100_000 cycles w/o crash and strange output only in one mode.
I modified the test code such that the increasing numbers are hand-coded into a bytearray, and sys.stdout.write() is used instead of print.
- thread0 REPL, thread1 burst(): No crash, no strange output. stopped at 100_00 cycles
- thread0 count_count(), thread1 burst(): No crash, char doubling about every 1000th message, stopped at 50_000 cycles
- thread0 burst, thread 1 count_count():  No crash, no strange output. stopped at 50_00 cycles

**Comment by robert-hh** [comment]
The character doubling in UART only occurs, if data is sent to stdout at the same time in thread 0. It does not matter if doing so by calling print() or by sys.stdout.write(). If I just let the LED blink, or interleave writing to stdout and to UART, everything works well.
interestingly, this character doubling only happens, if UART write is done in thread 1 and stdout write in thread 0. If the threads are swapped, it's fine.

**Comment by robert-hh** [comment]
I tried to figure out, whether the twice sent characters weer actually put into the UART FIFO by the code in uart_fill_tx_fifo(). So I modified the message to not set the same character twice in sequence, and checked in the uart_fill_tx_fifo() function, whether the actual character is the same as the previous one (code below). If yes, it would create a short pulse at a GPIO pin, which I then would be able to detect. This trigger happened much more rare that the character doubling. And that puzzles me. In a sequence of 100_000 messages, I had 74 with character duplications, but only one trigger event. b.t.w.: If the trigger happens, the function uart_fill_tx_fifo() was called in a interrupt context.

Modified uart_fill_tx_fifo(). P11 was set in the Python code as output. If the pulse is generated when two consecutive chars are different, it fires well.
```
uint32_t *PIO_OUT = (uint32_t *)0xd0000010;

// take bytes from the buffer and put them into the UART FIFO
STATIC void uart_fill_tx_fifo(machine_uart_obj_t *self) {
    static int last_char = 0;
    while (uart_is_writable(self->uart) && ringbuf_avail(&self->write_buffer) > 0) {
        // get a byte from the buffer and put it into the uart
        int c = ringbuf_get(&(self->write_buffer));
        if (c == last_char) {
            PIO_OUT[0] |= 1 << 11; // set P11
        }
        last_char = c;
        uart_get_hw(self->uart)->dr = c;
        PIO_OUT[0] &= ~(1 << 11);  // clear p11
    }
}
```

**Comment by peterhinch** [comment]
> That all looks like the order of execution is disturbed.

A high priority interrupt? Can recursive execution of the same ISR occur?

**Comment by robert-hh** [comment]
I thought about stacked ISR calls. But that should affect the sequence of bytes in the output. But he order is unchanged, just some bytes are sent twice in a row.

**Comment by robert-hh** [comment]
After increasing the output rate in thread0, the pulse width of the trigger pulse created in thread 1 is back to ~60ns. Very strange. And sometimes the ISR seems to last for more than 15ms, falling back to its usual ~20µs soon later. So it seems there is a lot of trouble with ISR and threading.

**Comment by robert-hh** [comment]
**Partially solved.** After having put aside this problem for a while, I had some progress now. If UART instatiation and UART use happen in the same thread, then the character doubling in the output does not happen.

The crash/lock-up which occurs when memory allocations are present is still present.

**Comment by dpgeorge** [comment]
The fix #7217 may be related to this, but not quite sure yet.  If you have a chance to test that PR with the issue here that would be helpful.

**Comment by robert-hh** [comment]
Sorry, it does not fix either of the two problems, I see here. It still stops after a while, if memory is allocated in the second thread. And if not, it still send character doubles if uart IRQ handler and uart send are not located in the same thread.

**Comment by jrullan** [comment]
I too have experienced similar issues in my project.  I documented my findings in the raspberry pi pico forum. 
https://forums.raspberrypi.com/viewtopic.php?t=324136

**Comment by peterhinch** [comment]
I wonder if https://github.com/micropython/micropython/issues/7977 is relevant. Interestingly that problem isn't restricted to UART output: SPI behaves identically.

**Comment by jrullan** [comment]
I'll repost here what I posted in the Raspberry Pi Pico forum in the hopes that this issue might be taken over again. The project I am attaching **does work**, but with several hacks and workarounds, none of which are fully characterized. The most strange/bizarre issue is the one requiring the "padding" array, look in the code below for MEMORY_ALIGNMENT_ARRAY.

> The culprit seems to be the implementation of the _thread module for the RP2. I noticed several things.
> 
> First it seems that the memory allocation done for the second thread somehow causes the garbage collector to stop working > unless specifically told to by using gc.collect() at specific places in the code, where memory leaks if and when the second thread is started.
> 
> Second, it seems that there is an issue of overlapping memory areas when the second thread is started because in order to make my code work I had to declare a block of memory in the form of an array of integers just before starting the second thread or otherwise the code stalls. Declaring the same array after starting the thread does not work and also the amount of integers is significant. I managed to make it work with any quantity above 512 integers which I think is related to the two 256 arrays I use to manipulate the LEDs.
> 
> Third, all the variables needed to be declared globally to prevent memory leakages.
> 
> I found a couple of posts on the Micropython forums that describe similar issues, so it seems the rp2 implementation of Micropython still needs work to iron out the memory issues.
> 
> Some links that have similar issues:
> 
> https://forum.micropython.org/viewtopic ... 21&t=10331
> https://forum.micropython.org/viewtopic ... lit=thread
> https://forum.micropython.org/viewtopic ... ead#p62348
> https://github.com/micropython/micropython/issues/7124

This is the code I used in my project, maybe it is useful to replicate and debug the problem. 

``` python
import array, utime, _thread, gc
from machine import Pin, UART
import rp2

try: 
    import gfx
except ImportError:
    raise RuntimeError("Cannot find gfx. Have you copied gfx.py to your Pico?")

#================
#   Objects
#================

# Serial object for the HC-05 communications
uart = UART(0,9600)

# Graphics object used to draw primitives
def set_pixel_xy(x,y,color,brightness=1):
    # Set a pixel using its x,y position (inverting row, column)
    set_pixel_rc(y,x,color,brightness)
    
graphics = gfx.GFX(32,8,set_pixel_xy)


#================
#   Constants
#================

# Matrix Configuration
STRIPS = 1
SUBMATRIX_HEIGHT = 8
SUBMATRIX_WIDTH = 32

# Configure the number of WS2812 LEDs.
NUM_LEDS = 256
PIN_NUM = 15

MAX_CYCLES = 5

#================
# Global Variables
#================
brightness = 0.1
color0 = 0
color1 = 0
color2 = 0
r=0
g=0
b=0
color_index = 0
color_cycles = 1

array_index = 0
s = 0
h = 0
p = 0

pixel_index = 0

alternate_colors = False
go = True

BLACK = (0, 0, 0)
RED = (255, 0, 0)
YELLOW = (255, 150, 0)
GREEN = (0, 255, 0)
CYAN = (0, 255, 255)
BLUE = (0, 0, 255)
PURPLE = (180, 0, 255)
WHITE = (255, 255, 255)
COLORS = (BLACK, RED, YELLOW, GREEN, CYAN, BLUE, PURPLE, WHITE)

new_color = COLORS[1]
color2 = COLORS[2]

# Create an array of 256 LED RGB values with initial 0 values.
ar = array.array("I", [0 for _ in range(NUM_LEDS)])
# Create a second array that will have the final values after adjusting for brightness
dimmer_ar = array.array("I", [0 for _ in range(NUM_LEDS)])


##########################################################################
# PIO WS2812b driver and state machine configuration
##########################################################################
@rp2.asm_pio(sideset_init=rp2.PIO.OUT_LOW, out_shiftdir=rp2.PIO.SHIFT_LEFT, autopull=True, pull_thresh=24)
def ws2812():
    T1 = 2
    T2 = 5
    T3 = 3
    wrap_target()
    label("bitloop")
    out(x, 1)               .side(0)    [T3 - 1] # Fetch one (1) bit from the OSR (the leds data), also set the pin to 0, also wait 2 extra cycles (3)
    jmp(not_x, "do_zero")   .side(1)    [T1 - 1] # If bit is 0, jump to "do_zero", also set pin to 1, also wait 1 extra cycle (2)
    jmp("bitloop")          .side(1)    [T2 - 1] # else fetch next bit, also set pin to 1, also wait 4 extra cycles (5)
    label("do_zero")
    nop()                   .side(0)    [T2 - 1] # do nothing, set pin to 0, wait 4 extra cycles (5)
    wrap()                                       # go back to wrap_target() without consuming any cycles (better than jmp)
sm = rp2.StateMachine(0, ws2812, freq=8_000_000, sideset_base=Pin(PIN_NUM))
sm.active(1)
##########################################################################


##########################################################################
# Basic pixels manipulation functions
##########################################################################
# Set a pixel using its index within the array
def pixels_set(i, color, brightness=1):
    global r
    global g
    global b
    r = int(color[0]*brightness)
    g = int(color[1]*brightness)
    b = int(color[2]*brightness)
    ar[i] = (g<<16) + (r<<8) + b

# Convert a row,column address to an index address
def rc_to_index(row,column):
    global array_index
    global s
    global h
    global p
    array_index = 0
    s = int(row/8)
    h = SUBMATRIX_HEIGHT
    p = column & 1
    if p==0:
        array_index = h*column + (row-h*s)
    else:
        array_index = h*column - (row-h*s) + 7
    array_index = STRIPS * array_index + s
    return array_index

# Set a pixel using row,column addressing
def set_pixel_rc(r,c,color,brightness=1):
    if r>=0 and c>=0 and r<SUBMATRIX_HEIGHT and c<SUBMATRIX_WIDTH:
        pixels_set(rc_to_index(r,c),color,brightness)

# Fill all pixels with color
def pixels_fill(color):
    for i in range(len(ar)):
        pixels_set(i, color)


# Apply brightness value to pixel colors and send to PIO state machine TX FIFO
def pixels_show():
    global r
    global g
    global b
    global pixel_index
    
    gc.collect() #<----------- Needed or memory leaking occurs (Only when the second thread is started)
    
    pixel_index = 0
    for pixel_index in range(NUM_LEDS):
        r = ((ar[pixel_index] >> 8) & 0xFF) * brightness
        g = ((ar[pixel_index] >> 16) & 0xFF) * brightness
        b = (ar[pixel_index] & 0xFF) * brightness
        dimmer_ar[pixel_index] = (int(g)<<16) + (int(r)<<8) + int(b)
    sm.put(dimmer_ar, 8)
      

##########################################################################
# Color changing functions    
##########################################################################
def switch_colors():
    global color_cycles
    global color_index

    color_cycles += 1
    if color_cycles > MAX_CYCLES:
        color_index += 1
        color_cycles = 1
    
    if (color_index >= len(COLORS)):
        color_index = 0
    
    return COLORS[color_index]

def swap_colors(color1,color2):
    global r
    global g
    global b
    r = (int((color2[0] - color1[0]) / (MAX_CYCLES - 1)) + color1[0]) if (color_cycles < MAX_CYCLES) else (color2[0])
    g = (int((color2[1] - color1[1]) / (MAX_CYCLES - 1)) + color1[1]) if (color_cycles < MAX_CYCLES) else (color2[1])
    b = (int((color2[2] - color1[2]) / (MAX_CYCLES - 1)) + color1[2]) if (color_cycles < MAX_CYCLES) else (color2[2])
    return (r, g, b)


##########################################################################
# Application specific drawing functions
##########################################################################
def faded_expanding_circle(x0,y0,r0,r1):
    for r in range(r0,r1+5):
        
        baton.acquire()  #<------- The lock is needed or it hangs when changing colors
        
        pixels_fill(BLACK)
        graphics.circle(x0,y0,r,new_color)
        if(r > r0):
            graphics.circle(x0,y0,r-1,new_color,0.3)        
        if(r > r0+1):
            graphics.circle(x0,y0,r-2,new_color,0.2)  
        if(r > r0+2):
            graphics.circle(x0,y0,r-3,new_color,0.08)  
        if(r > r0+3):
            graphics.circle(x0,y0,r-4,new_color,0.04)  
        pixels_show()
        
        baton.release()


#============================================================
# Bluetooth controls functions
#============================================================
# Checks if both the command string and the read buffer are equal
def equal(in_buf, cmd_str):
    cmd_len = len(cmd_str)
    
    for i in range(cmd_len):
        if cmd_str[i] != chr(in_buf[i]):
            return False
        
    return True

# Function that runs on the second core and handles bluetooth communications
def second_thread():
    global brightness
    global go
    global alternate_colors
    global new_color
    global color2
    global color_index
    global COLORS
    
    print("Second thread started")
    while True:

        if go and alternate_colors:
            
            baton.acquire()  #<--------- Needed to prevent lockups
            
            if color_index+1 >= len(COLORS):
                color2 = COLORS[1]
            else:
                color2 = COLORS[color_index+1]
            new_color = swap_colors(new_color,color2)
            switch_colors()
            
            baton.release()
        
        if uart.any():           
            print("messaged received")
            in_buf = uart.read()

            if equal(in_buf,"on"):
                go = True
                uart.write(str("PRENDIDAS","UTF-8")+chr(13)+chr(10))
                print("On")
            
            if equal(in_buf,"off"):
                uart.write(str("APAGADAS","UTF-8")+chr(13)+chr(10))
                go = False
                print("Off")
                
            if equal(in_buf, "red"):
                color_index = 1
                new_color = RED
                alternate_colors = False
                uart.write(str("RED","UTF-8")+chr(13)+chr(10))
            
            if equal(in_buf, "green"):
                color_index = 3
                new_color = GREEN
                alternate_colors = False
                uart.write(str("GREEN","UTF-8")+chr(13)+chr(10))
            
            if equal(in_buf, "blue"):
                color_index = 5
                new_color = BLUE
                alternate_colors = False
                uart.write(str("BLUE","UTF-8")+chr(13)+chr(10))

            if equal(in_buf, "alt"):
                if alternate_colors:
                    alternate_colors = False
                    uart.write(str("ALT OFF","UTF-8")+chr(13)+chr(10))
                else:
                    alternate_colors = True
                    uart.write(str("ALT ON","UTF-8")+chr(13)+chr(10))

            if equal(in_buf,"+"):
                brightness = brightness + 0.1
                if brightness > 1.0:
                    brightness = 1.0
                str_brightness = str(int(brightness*100)) + "%"
                print(str_brightness.encode())
                uart.write(str_brightness.encode()+chr(13)+chr(10))

            if equal(in_buf,"-"):
                brightness = brightness - 0.1
                if brightness < 0.0:
                    brightness = 0
                str_brightness = str(int(brightness*100)) + "%"
                print(str_brightness.encode())
                uart.write(str_brightness.encode()+chr(13)+chr(10))

            gc.collect()  #<------ Needed otherwise everytime a message arrives it leaks memory


#-----------------------------------------------------------------------------
#               MUST BE A MEMORY ALIGNMENT ISSUE!!!!!
#
#  The most bizarre hack! Without it the code hangs after just a few iterations
#  The number of values in the array that makes it work was found by trial and error
#  but it is my guess that the number has to be related to the two arrays used to
#  store the LEDs values, which take 256 values each.
#
#  It makes me hypothesize that there is an issue with memory allocation when using
#  the _thread module. This is not necessary when the second thread is not started.
#  Ah, btw it has to be created here, just before starting the thread. If it is
#  declared after starting the thread it does not work.
# 
MEMORY_ALIGNMENT_ARRAY = array.array("I", [0 for _ in range(513)])
#
#-----------------------------------------------------------------------------



#============================================================
# Main program
#============================================================
_thread.start_new_thread(second_thread, ())
baton = _thread.allocate_lock()

last_go = not go

def main():
    global last_go
    if go:        
        print("Color cycle",color_cycles, ", Color index ", color_index, " - ", new_color[0], new_color[1], new_color[2], gc.mem_free())
        faded_expanding_circle(15,3,0,16)
        if not last_go:
            last_go = True
    else:
        if last_go:
            pixels_fill(BLACK)
            pixels_show()
            last_go = False
            

while True:
    main()

```

This is the gfx library I used...

``` python

# Port of Adafruit GFX Arduino library to MicroPython.
# Based on: https://github.com/adafruit/Adafruit-GFX-Library
# Author: Tony DiCola (original GFX author Phil Burgess)
# License: MIT License (https://opensource.org/licenses/MIT)


class GFX:

    def __init__(self, width, height, pixel, hline=None, vline=None):
        # Create an instance of the GFX drawing class.  You must pass in the
        # following parameters:
        #  - width = The width of the drawing area in pixels.
        #  - height = The height of the drawing area in pixels.
        #  - pixel = A function to call when a pixel is drawn on the display.
        #            This function should take at least an x and y position
        #            and then any number of optional color or other parameters.
        #  You can also provide the following optional keyword argument to
        #  improve the performance of drawing:
        #  - hline = A function to quickly draw a horizontal line on the display.
        #            This should take at least an x, y, and width parameter and
        #            any number of optional color or other parameters.
        #  - vline = A function to quickly draw a vertical line on the display.
        #            This should take at least an x, y, and height paraemter and
        #            any number of optional color or other parameters.
        self.width = width
        self.height = height
        self._pixel = pixel
        # Default to slow horizontal & vertical line implementations if no
        # faster versions are provided.
        if hline is None:
            self.hline = self._slow_hline
        else:
            self.hline = hline
        if vline is None:
            self.vline = self._slow_vline
        else:
            self.vline = vline

    def _slow_hline(self, x0, y0, width, *args, **kwargs):
        # Slow implementation of a horizontal line using pixel drawing.
        # This is used as the default horizontal line if no faster override
        # is provided.
        if y0 < 0 or y0 > self.height or x0 < -width or x0 > self.width:
            return
        for i in range(width):
            self._pixel(x0+i, y0, *args, **kwargs)

    def _slow_vline(self, x0, y0, height, *args, **kwargs):
        # Slow implementation of a vertical line using pixel drawing.
        # This is used as the default vertical line if no faster override
        # is provided.
        if y0 < -height or y0 > self.height or x0 < 0 or x0 > self.width:
            return
        for i in range(height):
            self._pixel(x0, y0+i, *args, **kwargs)

    def rect(self, x0, y0, width, height, *args, **kwargs):
        # Rectangle drawing function.  Will draw a single pixel wide rectangle
        # starting in the upper left x0, y0 position and width, height pixels in
        # size.
        if y0 < -height or y0 > self.height or x0 < -width or x0 > self.width:
            return
        self.hline(x0, y0, width, *args, **kwargs)
        self.hline(x0, y0+height-1, width, *args, **kwargs)
        self.vline(x0, y0, height, *args, **kwargs)
        self.vline(x0+width-1, y0, height, *args, **kwargs)

    def fill_rect(self, x0, y0, width, height, *args, **kwargs):
        # Filled rectangle drawing function.  Will draw a single pixel wide
        # rectangle starting in the upper left x0, y0 position and width, height
        # pixels in size.
        if y0 < -height or y0 > self.height or x0 < -width or x0 > self.width:
            return
        for i in range(x0, x0+width):
            self.vline(i, y0, height, *args, **kwargs)

    def line(self, x0, y0, x1, y1, *args, **kwargs):
        # Line drawing function.  Will draw a single pixel wide line starting at
        # x0, y0 and ending at x1, y1.
        steep = abs(y1 - y0) > abs(x1 - x0)
        if steep:
            x0, y0 = y0, x0
            x1, y1 = y1, x1
        if x0 > x1:
            x0, x1 = x1, x0
            y0, y1 = y1, y0
        dx = x1 - x0
        dy = abs(y1 - y0)
        err = dx // 2
        ystep = 0
        if y0 < y1:
            ystep = 1
        else:
            ystep = -1
        while x0 <= x1:
            if steep:
                self._pixel(y0, x0, *args, **kwargs)
            else:
                self._pixel(x0, y0, *args, **kwargs)
            err -= dy
            if err < 0:
                y0 += ystep
                err += dx
            x0 += 1

    def circle(self, x0, y0, radius, *args, **kwargs):
        # Circle drawing function.  Will draw a single pixel wide circle with
        # center at x0, y0 and the specified radius.
        f = 1 - radius
        ddF_x = 1
        ddF_y = -2 * radius
        x = 0
        y = radius
        self._pixel(x0, y0 + radius, *args, **kwargs)
        self._pixel(x0, y0 - radius, *args, **kwargs)
        self._pixel(x0 + radius, y0, *args, **kwargs)
        self._pixel(x0 - radius, y0, *args, **kwargs)
        while x < y:
            if f >= 0:
                y -= 1
                ddF_y += 2
                f += ddF_y
            x += 1
            ddF_x += 2
            f += ddF_x
            self._pixel(x0 + x, y0 + y, *args, **kwargs)
            self._pixel(x0 - x, y0 + y, *args, **kwargs)
            self._pixel(x0 + x, y0 - y, *args, **kwargs)
            self._pixel(x0 - x, y0 - y, *args, **kwargs)
            self._pixel(x0 + y, y0 + x, *args, **kwargs)
            self._pixel(x0 - y, y0 + x, *args, **kwargs)
            self._pixel(x0 + y, y0 - x, *args, **kwargs)
            self._pixel(x0 - y, y0 - x, *args, **kwargs)

    def fill_circle(self, x0, y0, radius, *args, **kwargs):
        # Filled circle drawing function.  Will draw a filled circule with
        # center at x0, y0 and the specified radius.
        self.vline(x0, y0 - radius, 2*radius + 1, *args, **kwargs)
        f = 1 - radius
        ddF_x = 1
        ddF_y = -2 * radius
        x = 0
        y = radius
        while x < y:
            if f >= 0:
                y -= 1
                ddF_y += 2
                f += ddF_y
            x += 1
            ddF_x += 2
            f += ddF_x
            self.vline(x0 + x, y0 - y, 2*y + 1, *args, **kwargs)
            self.vline(x0 + y, y0 - x, 2*x + 1, *args, **kwargs)
            self.vline(x0 - x, y0 - y, 2*y + 1, *args, **kwargs)
            self.vline(x0 - y, y0 - x, 2*x + 1, *args, **kwargs)

    def triangle(self, x0, y0, x1, y1, x2, y2, *args, **kwargs):
        # Triangle drawing function.  Will draw a single pixel wide triangle
        # around the points (x0, y0), (x1, y1), and (x2, y2).
        self.line(x0, y0, x1, y1, *args, **kwargs)
        self.line(x1, y1, x2, y2, *args, **kwargs)
        self.line(x2, y2, x0, y0, *args, **kwargs)

    def fill_triangle(self, x0, y0, x1, y1, x2, y2, *args, **kwargs):
        # Filled triangle drawing function.  Will draw a filled triangle around
        # the points (x0, y0), (x1, y1), and (x2, y2).
        if y0 > y1:
            y0, y1 = y1, y0
            x0, x1 = x1, x0
        if y1 > y2:
            y2, y1 = y1, y2
            x2, x1 = x1, x2
        if y0 > y1:
            y0, y1 = y1, y0
            x0, x1 = x1, x0
        a = 0
        b = 0
        y = 0
        last = 0
        if y0 == y2:
            a = x0
            b = x0
            if x1 < a:
                a = x1
            elif x1 > b:
                b = x1
            if x2 < a:
                a = x2
            elif x2 > b:
                b = x2
            self.hline(a, y0, b-a+1, *args, **kwargs)
            return
        dx01 = x1 - x0
        dy01 = y1 - y0
        dx02 = x2 - x0
        dy02 = y2 - y0
        dx12 = x2 - x1
        dy12 = y2 - y1
        if dy01 == 0:
            dy01 = 1
        if dy02 == 0:
            dy02 = 1
        if dy12 == 0:
            dy12 = 1
        sa = 0
        sb = 0
        if y1 == y2:
            last = y1
        else:
            last = y1-1
        for y in range(y0, last+1):
            a = x0 + sa // dy01
            b = x0 + sb // dy02
            sa += dx01
            sb += dx02
            if a > b:
                a, b = b, a
            self.hline(a, y, b-a+1, *args, **kwargs)
        sa = dx12 * (y - y1)
        sb = dx02 * (y - y0)
        while y <= y2:
            a = x1 + sa // dy12
            b = x0 + sb // dy02
            sa += dx12
            sb += dx02
            if a > b:
                a, b = b, a
            self.hline(a, y, b-a+1, *args, **kwargs)
            y += 1
```

**Comment by helarsen** [comment]
Allow me to piggy back onto this thread with similar findings of corruption when using _thread and UART on RPi Pico. Reported here.
[https://forums.raspberrypi.com/viewtopic.php?t=329639](url)
I can provide the complete code if it makes sense.

**Comment by yjchun** [comment]
I was testing this issue and found something interesting.
With following two modifications, I could run above test code successfully (at least for 30 minutes).

1. modify ports/rp2/mpthreadport.c:59
```python
            //gc_collect_root((void **)core1_stack, core1_stack_num_words);
            gc_collect_root((void **)&core1_stack, core1_stack_num_words);
```

2. main.py: put gc.collect() in main loop.
```python
from machine import UART, Pin
import time
import _thread
import gc

uart = UART(0, 115200 , parity=None, stop=1, bits=8, tx=Pin(0), rx=Pin(1), txbuf=32, timeout=10)

def send(command):
        write_buffer = bytearray(len(command) + 3)
        write_buffer[0: len(command)] = bytes(command, 'ASCII')
        write_buffer[-3:] = b'\xff\xff\xff'
        uart.write(write_buffer)

def burst():
    msg="0123456789" * 8 + "\r\n"
    n=0
    while True:
        uart.write("%d: " % n)
        uart.write(msg)
        n += 1

def count_count():
    n = 1
    while True:
        print(gc.mem_free())
        gc.collect()
        print("Count Count counts", n)
        n += 1
        time.sleep(1)

_thread.start_new_thread(burst, ())
count_count()
```
With these two modifications, it ran successfully.
My test results: 
1. I couldn't explain why this works reading the code: `gc_collect_root((void **)&core1_stack, core1_stack_num_words);`
2. without gc.collect() in main thread, the moment GC is out of memory is when thread stops.
3. gc.collect() in thread stop the thread, same reason with (2)
4. I did notice character doubling problem very rarely in test cases. I am using master rev. as of today. Instead, without modification (1), uart is sending some garbages when burst() is called from main thread: `930: €Rå0 931: €RdP 932: €Rçp`. Only heap data is lost.

After tests, my guess is that GC information (or address) is not properly setup in `core1`. I will try to look into it later but I am not familiar with micropython source code yet.

**Comment by yjchun** [comment]
A quick note on character doubling problem: When I was reading rp2/machine_uart.c, It was suspicious that write_buffer and read_buffer is not properly protected against IRQ: write_lock and read_lock to be set for all access to read/write_buffer? This might be the cause but didn't verify yet though.

**Comment by dpgeorge** [comment]
Fixed by 2d47020e15b82b7f021e0b52dabe75baf5cb4c4d
